src/System.Memory/tests/ArrayBufferWriter/ArrayBufferWriterTests.String.cs

```diff
@@ -0,0 +1,315 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+using Xunit;
+
+namespace System.Buffers.Tests
+{
+    public static partial class ArrayBufferWriterTests_String
```

Comment:
I didn't realize we want to support this on non-primitive types; string is normally immutable as well.

Reply:
> I didn't realize we want to support this on non-primitive types

This is a question on the interface, tbh: https://github.com/dotnet/corefx/blob/master/src/System.Memory/src/System/Buffers/IBufferWriter.cs

The test using string is just to verify that T : reference type works fine. As you mentioned, strings are immutable, so I don't expect anyone to create an `ABW<string>`.

**I can remove this test all together.**

It doesn't have any constraints on T. Maybe it should be T: struct or T: unmanaged.

cc @davidfowl, @stephentoub

Reply:
Also cc @jkotas, if you have any thoughts on the issue.

Reply:
`IBufferWriter` shipped without any constraints. Do we consider it to be a mistake?

Since `IBufferWriter` does not have any constraints, `ArrayBufferWriter` should not have any constraints either for consistency?